### PR TITLE
Add support for PostgreSQL exclusion constraints

### PR DESCRIPTION
**Pull request by eoin-obrien**
I've added an `@Exclusion` entity-level decorator patterned after `@Check` and `@Unique` to support [PostgreSQL exclusion constraints](https://www.postgresql.org/docs/current/static/ddl-constraints.html#DDL-CONSTRAINTS-EXCLUSION) as described in #2698.

I've also added the necessary test cases to cover exclusion constraints, and documented the new decorator in `decorator-reference.md`.

Example usage:

```typescript
@Entity()
@Exclusion(`USING gist ("room" WITH =, tsrange("from", "to") WITH &&)`)
export class RoomBooking {

    @Column()
    room: string;

    @Column()
    from: Date;

    @Column()
    to: Date;
}
```

### CODE REVIEW

File: src/driver/postgres/PostgresQueryRunner.ts [hunk 1]
```diff
@@ -1241,12 +1289,13 @@ export class PostgresQueryRunner extends BaseQueryRunner implements QueryRunner
...
-            `CASE "cnst"."contype" WHEN 'p' THEN 'PRIMARY' WHEN 'u' THEN 'UNIQUE' WHEN 'c' THEN 'CHECK' END AS "constraint_type", "a"."attname" AS "column_name" ` +
+        const constraintsSql = `SELECT "ns"."nspname" AS "table_schema", "t"."relname" AS "table_name", "cnst"."conname" AS "constraint_name", ` +
+            `CASE "cnst"."contype" WHEN 'x' THEN pg_get_constraintdef("cnst"."oid", true) ELSE "cnst"."consrc" END AS "expression", ` +
+            `CASE "cnst"."contype" WHEN 'p' THEN 'PRIMARY' WHEN 'u' THEN 'UNIQUE' WHEN 'c' THEN 'CHECK' WHEN 'x' THEN 'EXCLUDE' END AS "constraint_type", "a"."attname" AS "column_name" ` +
             `FROM "pg_constraint" "cnst" ` +
             `INNER JOIN "pg_class" "t" ON "t"."oid" = "cnst"."conrelid" ` +
             `INNER JOIN "pg_namespace" "ns" ON "ns"."oid" = "cnst"."connamespace" ` +
-            `INNER JOIN "pg_attribute" "a" ON "a"."attrelid" = "cnst"."conrelid" AND "a"."attnum" = ANY ("cnst"."conkey") ` +
+            `LEFT JOIN "pg_attribute" "a" ON "a"."attrelid" = "cnst"."conrelid" AND "a"."attnum" = ANY ("cnst"."conkey") ` +
```
Comment (pleerock) [inline src/driver/postgres/PostgresQueryRunner.ts]: can you please explain why did you change INNER to LEFT?

**Comment by pleerock** [comment]
Hey, sorry for the late response here.

Thank you very much, this is huge work.
But I'm really scary to merge such feature requests in such responsible places 😨 Are you sure you check everything? (yeah sorry for dumb question, but can you please check recheck everything again to make sure everything is alright :))

File: src/driver/postgres/PostgresQueryRunner.ts (hunk 1 above)
Comment (eoin-obrien) [reply src/driver/postgres/PostgresQueryRunner.ts]: Sure! From the PostgreSQL docs on [`pg_constraint`](https://www.postgresql.org/docs/9.0/static/catalog-pg-constraint.html):

> In the case of an exclusion constraint, `conkey` is only useful for constraint elements that are simple column references. For other cases, a zero appears in `conkey` and the associated index must be consulted to discover the expression that is constrained. (`conkey` thus has the same contents as `pg_index.indkey` for the index.)

If the join is left as an `INNER JOIN`, then any exclusion constraints that aren't simple column references will be removed from the selected rows, as there won't be a related row in `pg_attribute` for them. Hence the `LEFT JOIN` to keep those rows in the selection!

I'm grabbing the constraint expression using `pg_get_constraintdef` instead, as per the docs link above.

**Comment by eoin-obrien** [comment]
No worries @pleerock – I went ahead and double-checked the changes there, as well as rebasing the PR from `master` to make sure there wouldn't be any merge conflicts. All of the test cases pass, insofar as I can tell! :+1: 

Most of the diff adds new lines of code, as opposed to editing existing lines. The `LEFT JOIN` section you marked in your review is one of the few sections where I actually had to make changes to a pre-existing feature.

**Comment by eoin-obrien** [comment]
@pleerock – anything I can do to help get this merged?

**Comment by pleerock** [comment]
No, lets merge it. Thank you very much!
